File: 0084 largest-rectangle-in-histogram/solution.java
class Solution {
    public int largestRectangleArea(int[] h) {
        int[] st=new int[h.length+1];
        int max=0,top=-1;
        int[] res = new int[h.length];
        for(int i=0;i<=h.length;i++) {
            int he=(i==h.length ? 0 : h[i]);
            while(top!=-1 && he < h[st[top]]) {
                int height=h[st[top--]];
                int width = (top==-1 ? i : i-st[top]-1);
                max=Math.max(max,height*width);
            }
            st[++top]=i;
        }
        return max;
    }
}

File: 0084 largest-rectangle-in-histogram/solution.go
func largestRectangleArea(h []int) int {
    st:=make([]int,len(h)+1)
    top,ma:=-1,0
    for i:=0;i<=len(h);i++ {
        he:=0
        if i!=len(h) {
            he=h[i]
        }
        for top!=-1 && h[st[top]]>he {
            height:=h[st[top]]
            top--
            width:= i
            if top!=-1 {
                width=i-st[top]-1
            }
            ma=max(ma,height*width)
        }
        top++
        st[top]=i
    }
    return ma
}
